controllers/storagecluster/clusterclaims.go

```diff
@@ -0,0 +1,226 @@
+
+	for _, csv := range csvs.Items {
+		if strings.HasPrefix(csv.Name, OdfOperatorNamePrefix) {
+			return csv.Name, nil
+		}
+	}
+
+	return "", fmt.Errorf("failed to find csv with prefix %q", OdfOperatorNamePrefix)
+}
+
+func (c *ClusterClaimCreator) getCephFsid() (string, error) {
+	var rookCephMonSecret corev1.Secret
+	err := c.Client.Get(c.Context, types.NamespacedName{Name: RookCephMonSecretName, Namespace: c.StorageCluster.Namespace}, &rookCephMonSecret)
+	if err != nil {
+		return "", err
+	}
+	if val, ok := rookCephMonSecret.Data[FsidKey]; ok {
+		return string(val), nil
+	}
+
+	return "", fmt.Errorf("failed to fetch ceph fsid from %q secret", RookCephMonSecretName)
+}
+
+func (c *ClusterClaimCreator) setStorageSystemName(name string) *ClusterClaimCreator {
+	c.Values[StorageSystemName] = name
+	return c
+}
+
+func (c *ClusterClaimCreator) setOdfVersion(version string) *ClusterClaimCreator {
+	c.Values[OdfVersion] = version
+	return c
+}
+
+func (c *ClusterClaimCreator) setStorageClusterName(name string) *ClusterClaimCreator {
+	c.Values[StorageClusterName] = name
+	return c
+}
+
+func (c *ClusterClaimCreator) setCephFsid(fsid string) *ClusterClaimCreator {
+	c.Values[CephFsid] = fsid
+	return c
+}
+
+func (c *ClusterClaimCreator) getStorageSystemName() (string, error) {
+	for _, ref := range c.StorageCluster.OwnerReferences {
+		if ref.Kind == "StorageSystem" {
+			return ref.Name, nil
+		}
+	}
+
+	return "", fmt.Errorf("failed to find parent StorageSystem's name in StorageCluster %q ownerreferences", c.StorageCluster.Name)
+}
+
+func (obj *ocsClusterClaim) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	r.Log.Info("deleting ClusterClaim resources")
+	ctx := context.TODO()
+	if crdExists, err := checkIfClusterClaimCRDExists(ctx, r.Client); !crdExists {
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+	names := []string{OdfVersion, StorageSystemName, StorageClusterName, CephFsid}
+	for _, name := range names {
+		cc := clusterv1alpha1.ClusterClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+		}
+
+		var foundClusterClaim clusterv1alpha1.ClusterClaim
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: cc.Name}, &foundClusterClaim)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				r.Log.Info("ClusterClaim %q not found.", cc.Name)
+				continue
+			}
+			return reconcile.Result{}, fmt.Errorf("unable to retrieve ClusterClaim %v: %v", cc.Name, err)
+		}
+
+		if foundClusterClaim.GetDeletionTimestamp().IsZero() {
+			r.Log.Info("deleting ClusterClaim %q.", foundClusterClaim.Name)
+			err = r.Client.Delete(context.TODO(), &foundClusterClaim)
+			if err != nil {
+				r.Log.Error(err, "failed to delete ClusterClaim %q.", cc.Name)
+				return reconcile.Result{}, fmt.Errorf("failed to delete %v: %v", foundClusterClaim.Name, err)
+			}
+		}
+
+		r.Log.Info("successfully deleted ClusterClaim %q.", cc.Name)
+	}
```

Comment:
Can we just do delete calls instead of get and delete and ignore error if delete returns not found

Reply:
done but don't you think we should be checking if the latest version of resource exists first and if it is already marked for deletion ?

Reply:
Actually, IMO it will save a few client calls, earlier for every cluster claim we were making 2 client calls and now it is 1.